storage/badger.go

```diff
@@ -0,0 +1,230 @@
+	db            *badger.DB
+	opts          badger.Options
+	useWriteBatch bool
+}
+
+func NewBadgerDatabase(opts badger.Options, useWriteBatch bool) *BadgerDatabase {
+	return &BadgerDatabase{
+		db:            nil,
+		opts:          opts,
+		useWriteBatch: useWriteBatch,
+	}
+}
+
+func (bdb *BadgerDatabase) Setup() error {
+	db, err := badger.Open(bdb.opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+	bdb.db = db
+	return nil
+}
+
+func (bdb *BadgerDatabase) GetContext(id []byte) Context {
+	return NewBadgerContext(id, bdb.useWriteBatch)
+}
+
+func (bdb *BadgerDatabase) Update(ctx Context, fn func(Transaction, Context) error) error {
+	var wb *badger.WriteBatch
+	if bdb.useWriteBatch {
+		wb = bdb.db.NewWriteBatch()
+		defer wb.Cancel()
+		defer wb.Flush()
+	}
+
+	return bdb.db.Update(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, wb)
+		return fn(T, ctx)
+	})
+}
+
+func (bdb *BadgerDatabase) View(ctx Context, fn func(Transaction, Context) error) error {
+	return bdb.db.View(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, nil)
+		return fn(T, ctx)
+	})
+}
+
+func (bdb *BadgerDatabase) Close() error {
+	return bdb.db.Close()
+}
+
+func (bdb *BadgerDatabase) Erase() error {
+	return os.RemoveAll(bdb.opts.Dir)
+}
+
+func (bdb *BadgerDatabase) Id() DatabaseId {
+	return BADGERDB
+}
+
+// ==========================
+// BadgerTransaction
+// ==========================
+
+type BadgerTransaction struct {
+	txn *badger.Txn
+	wb  *badger.WriteBatch
+}
+
+func NewBadgerTransaction(txn *badger.Txn, wb *badger.WriteBatch) *BadgerTransaction {
+	return &BadgerTransaction{
+		txn: txn,
+		wb:  wb,
+	}
+}
+
+func (btx *BadgerTransaction) Set(key []byte, value []byte, ctx Context) error {
+	prefixedKey, err := castBadgerContextAndGetPrefixedKey(key, ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Set:")
+	}
+
+	if btx.wb != nil {
+		return btx.wb.Set(prefixedKey, value)
+	}
+	return btx.txn.Set(prefixedKey, value)
+}
+
+func (btx *BadgerTransaction) Delete(key []byte, ctx Context) error {
+	prefixedKey, err := castBadgerContextAndGetPrefixedKey(key, ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Delete:")
+	}
+
+	if btx.wb != nil {
+		return btx.wb.Delete(prefixedKey)
+	}
+	return btx.txn.Delete(prefixedKey)
+}
+
+func (btx *BadgerTransaction) Get(key []byte, ctx Context) ([]byte, error) {
+	var value []byte
+	prefixedKey, err := castBadgerContextAndGetPrefixedKey(key, ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Get:")
+	}
+
+	item, err := btx.txn.Get(prefixedKey)
+	if err != nil {
+		return value, errors.Wrapf(err, "Get:")
+	}
+	return item.ValueCopy(nil)
+}
+
+func (btx *BadgerTransaction) GetIterator(ctx Context) (Iterator, error) {
+	badgerCtx, err := AssertContext[*BadgerContext](ctx, BADGERDB)
+	if err != nil {
+		return nil, err
+	}
+	opts := badger.DefaultIteratorOptions
+	it := btx.txn.NewIterator(opts)
+	it.Seek(badgerCtx.prefix)
+	return NewBadgerIterator(it, badgerCtx), nil
+}
```

Comment:
What's interesting about the `GetIterator(ctx)` interface function is that we generally always want to create a new iterator given a prefix and transaction. Golang's type system is also quite limited (no support for variance) which is partly why it forces you to implement the `AssertContext` function. 

My general suggestion for golang dev is to avoid having functions that require you to pass in interface types that you internally need to validate and cast.

As an alternative here, what you can do is update the constructor for NewBadgerIterator and have it incapsulate the logic for creating the iterator:
```suggestion
func NewBadgerIterator(txn *BadgerTransaction, prefix []byte) *BadgerIterator {
	opts := badger.DefaultIteratorOptions
	it := txn.NewIterator(opts)
	it.Seek(prefix)
	return &BadgerIterator{
		it:  it,
	}
}
```